File: leetcode/most-frequent-subtree-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
    int max = -1;
    public int[] findFrequentTreeSum(TreeNode root) {
        if(root == null){
            return new int[0];
        }
        populate_sums(root);
        ArrayList<Integer> result = new ArrayList<Integer>();
        for(Map.Entry<Integer, Integer> entry : map.entrySet()){
            if(entry.getValue() == max){
                result.add(entry.getKey());
            }
        }
        int[] r = new int[result.size()];
       
        for(int i = 0; i < r.length; i++){
            r[i] = result.get(i);
        }
        return r;
    }
    private int populate_sums(TreeNode node){
        int sum = 0;
        if(node.left != null){
            sum += populate_sums(node.left);
        }
        sum += node.val;
        if(node.right != null){
            sum += populate_sums(node.right);
        }
        //System.out.println(sum);
        if(map.containsKey(sum)){
            int i = map.get(sum)+1;
            map.put(sum, i);
            if(i > max){
                max = i;
            }
        }else{
            map.put(sum,1);
            if(1 > max){
                max = 1;
            }
        }
        return sum;
    }

File: leetcode/most-frequent-subtree-sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import defaultdict
class Solution(object):
    def findFrequentTreeSum(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def handleSum(sumAmount):
            self.frequencies[sumAmount] += 1
            if self.frequencies[sumAmount] == self.highest_freq:
                self.result.append(sumAmount)
            elif self.frequencies[sumAmount] > self.highest_freq:
                self.highest_freq = self.frequencies[sumAmount]
                self.result = [sumAmount]

        def evaluate(node):
            if node.left != None:
                evaluate(node.left)
                node.val += node.left.val
            if node.right != None:
                evaluate(node.right)
                node.val += node.right.val
            handleSum(node.val)

        self.highest_freq = 0
        self.frequencies = defaultdict(int)
        self.result = []
        if root == None:
            return []

        evaluate(root)
        return self.result